### PR TITLE
server xml variable and app property config support

### LINKED ISSUE #2643: MP Config 1.3: Liberty variables ConfigSource
This will access variables via the VariablesRegistry but need to exclude some system variables ... TBD

### PR DESCRIPTION
**Pull request by neilgsyoung**
fixes #2643

### CODE REVIEW

**Comment by neilgsyoung** [comment]
#build

File: dev/com.ibm.ws.microprofile.config.1.3/.classpath [hunk 1]
```diff
@@ -6,5 +6,8 @@
 	<classpathentry kind="src" output="bin_test" path="test/resources"/>
 	<classpathentry kind="con" path="aQute.bnd.classpath.container"/>
 	<classpathentry kind="con" path="org.eclipse.jdt.launching.JRE_CONTAINER/org.eclipse.jdt.internal.debug.ui.launcher.StandardVMType/JavaSE-1.8"/>
+	<classpathentry combineaccessrules="false" exported="true" kind="src" path="/com.ibm.ws.microprofile.config.1.2"/>
```
Comment (tevans78) [inline dev/com.ibm.ws.microprofile.config.1.3/.classpath]: You can't add these lines, should all be done via bnd

File: dev/com.ibm.ws.microprofile.config.1.3/bnd.bnd [hunk 2]
```diff
@@ -44,6 +44,7 @@ Import-Package: \
 WS-TraceGroup: APPCONFIG
 
 Include-Resource: \
+  OSGI-INF=resources/OSGI-INF,\
```
Comment (tevans78) [inline dev/com.ibm.ws.microprofile.config.1.3/bnd.bnd]: Could have sworn I removed this line :-o

File: dev/com.ibm.ws.microprofile.config.1.3/src/com/ibm/ws/microprofile/config13/sources/OSGiConfigUtils.java [hunk 3]
```diff
@@ -178,22 +190,40 @@ public static String getCDIAppName(BundleContext bundleContext) {
...
      * @return A filter string to be used by the OSGi ConfigurationAdmin service
      */
-    public static String getApplicationConfigFilter(BundleContext bundleContext, String applicationName) {
+    public static String getApplicationPropertiesConfigFilter(String applicationPid) {
         String applicationConfigFilter = null;
 
-        String applicationPID = getApplicationPID(bundleContext, applicationName);
+        StringBuilder applicationFilter = new StringBuilder(200);
+        applicationFilter.append("(&");
+        applicationFilter.append(FilterUtils.createPropertyFilter("service.factoryPid", "com.ibm.ws.appconfig.appProperties"));
```
Comment (tevans78) [inline dev/com.ibm.ws.microprofile.config.1.3/src/com/ibm/ws/microprofile/config13/sources/OSGiConfigUtils.java]: Ideally I think these strings should be extracted out into constants at the top of this class

File: dev/com.ibm.ws.microprofile.config.1.3/src/com/ibm/ws/microprofile/config13/sources/OSGiConfigUtils.java [hunk 4]
```diff
@@ -205,27 +235,60 @@ public static String getApplicationConfigFilter(BundleContext bundleContext, Str
...
 
+    /**
+     * Get a Map that represents the name/value pairs of <variable> elements in the server.xml
+     *
+     * @param bundleContext the context to use in looking up OSGi service references
+     * @return
+     */
+    public static Map<String, String> getVariableFromServerXML(BundleContext bundleContext) {
+
+        ConfigVariables configVars = getConfigVariables(bundleContext);
```
Comment (tevans78) [inline dev/com.ibm.ws.microprofile.config.1.3/src/com/ibm/ws/microprofile/config13/sources/OSGiConfigUtils.java]: Can/should `ConfigVariables` be cached?

File: dev/com.ibm.ws.microprofile.config.1.3_fat/fat/src/com/ibm/ws/microprofile/config13/test/VariableServerXMLTest.java [hunk 5]
```diff
@@ -0,0 +1,155 @@
...
+        // run the "before" test to check the value of the variable before the server.xml is updated
+        test(server, "/variableServerXMLApp/ServerXMLVariableServlet?testMethod=varPropertiesBeforeTest");
+
+        // switch to new configuration
+        server.copyFileToLibertyServerRoot("refreshVariables/server.xml");
+
+        // Wait for message: "the server configuration was successfully updated"
+        assertNotNull("The server.xml was not updated",
+                      server.waitForStringInLog("CWWKG0017I"));
+        Thread.sleep(1000); // We need this pause so that the config change is picked up through the polling mechanism,
```
Comment (tevans78) [inline dev/com.ibm.ws.microprofile.config.1.3_fat/fat/src/com/ibm/ws/microprofile/config13/test/VariableServerXMLTest.java]: if it takes time for the change in server.xml to be detected, shouldn't this wait be before we check for the string in the log?

File: dev/com.ibm.ws.microprofile.config.1.3_fat/fat/src/com/ibm/ws/microprofile/config13/test/VariableServerXMLTest.java [hunk 6]
```diff
@@ -0,0 +1,155 @@
...
+        // run the "before" test to check the value of the variable before the server.xml is updated
+        test(server, "/variableServerXMLApp/ServerXMLVariableServlet?testMethod=appPropertiesBeforeTest");
+
+        // switch to new configuration
+        server.copyFileToLibertyServerRoot("refreshAppProperties/server.xml");
+
+        // Wait for message: "the server configuration was successfully updated"
+        assertNotNull("The server.xml was not updated",
+                      server.waitForStringInLog("CWWKG0017I"));
+        Thread.sleep(1000); // We need this pause so that the config change is picked up through the polling mechanism,
```
Comment (tevans78) [inline dev/com.ibm.ws.microprofile.config.1.3_fat/fat/src/com/ibm/ws/microprofile/config13/test/VariableServerXMLTest.java]: ditto

File: dev/com.ibm.ws.microprofile.config.1.3_fat/publish/files/refreshVariables/server.xml [hunk 7]
```diff
@@ -0,0 +1,33 @@
+<server description="Server for testing variable Config">
```
Comment (tevans78) [inline dev/com.ibm.ws.microprofile.config.1.3_fat/publish/files/refreshVariables/server.xml]: is this server.xml the same as the previous one?

File: dev/com.ibm.ws.microprofile.config.1.3_fat/publish/servers/ServerXMLVariableServer/server.xml [hunk 8]
```diff
@@ -0,0 +1,33 @@
...
+		<feature>servlet-4.0</feature>
+		<feature>componentTest-1.0</feature>
+		<feature>mpConfig-1.3</feature>
+	</featureManager>
+
+    <variable name="string_variable" value="string_value"/>
+    <variable name="int_variable" value="999"/>
+    <variable name="bool_variable" value="true"/>
+    <variable name="multi_string_variable" value="first,second,third"/>
+    <variable name="changeable_variable" value="original_value" />
```
Comment (tevans78) [inline dev/com.ibm.ws.microprofile.config.1.3_fat/publish/servers/ServerXMLVariableServer/server.xml]: just this line which is different yeah?

File: dev/com.ibm.ws.microprofile.config.1.3/resources/OSGI-INF/l10n/metatype.properties [hunk 9]
```diff
@@ -19,3 +19,9 @@
 
 appProperties=Application scoped properties
 appProperties.desc=Provides custom properties to an application.
+property=A name/value pair describing a property
```
Comment (rhpeters) [inline dev/com.ibm.ws.microprofile.config.1.3/resources/OSGI-INF/l10n/metatype.properties]: ID reviewed

**Comment by neilgsyoung** [comment]
#build

File: dev/com.ibm.ws.microprofile.config.1.3_fat/fat/src/com/ibm/ws/microprofile/config13/test/VariableServerXMLTest.java (hunk 5 above)
Comment (neilgsyoung) [reply dev/com.ibm.ws.microprofile.config.1.3_fat/fat/src/com/ibm/ws/microprofile/config13/test/VariableServerXMLTest.java]: The waitForStringInLog() method waits for the server.xml to be refreshed and for the appropriate message to appear in the log. The sleep allows time for a delay in the mp config polling mechanism re-reading that (new) configuration.

File: dev/com.ibm.ws.microprofile.config.1.3_fat/fat/src/com/ibm/ws/microprofile/config13/test/VariableServerXMLTest.java (hunk 6 above)
Comment (neilgsyoung) [reply dev/com.ibm.ws.microprofile.config.1.3_fat/fat/src/com/ibm/ws/microprofile/config13/test/VariableServerXMLTest.java]: ditto my comment on your comment :-)

File: dev/com.ibm.ws.microprofile.config.1.3_fat/publish/servers/ServerXMLVariableServer/server.xml (hunk 8 above)
Comment (neilgsyoung) [reply dev/com.ibm.ws.microprofile.config.1.3_fat/publish/servers/ServerXMLVariableServer/server.xml]: yup

File: dev/com.ibm.ws.microprofile.config.1.3_fat/publish/files/refreshVariables/server.xml (hunk 7 above)
Comment (neilgsyoung) [reply dev/com.ibm.ws.microprofile.config.1.3_fat/publish/files/refreshVariables/server.xml]: I've tweaked it to make clear the distinction between testing a changed server.xml "variable" attribute and appProperties "property" element.

**Comment by neilgsyoung** [comment]
#build

**Comment by wraschke** [comment]
This pull request has been reverted because it caused a regression in the LAOS continuous build at

https://wasrtc.hursley.ibm.com:9443/jazz/resource/itemOid/com.ibm.team.build.BuildResult/_A09dsIQ7EeiADe06pEGqdQ

Test failures so far seen (the build is still running) are in:

com.ibm.ws.config.internal.schemaChecker.MetaTypeValidator
- com.ibm.ws.appconfig.appProperties elementLabelNotTitleCase
- com.ibm.ws.appconfig.appProperties pidLabelNotTitleCase
- com.ibm.ws.appconfig.appProperties.property pidLabelNotTitleCase

**Pull request merged**